_posts/2021-02-09-non-key-joins-in-kstreams.adoc

```diff
@@ -0,0 +1,288 @@
+----
+
+== Building and Running the Application
+
+You can now build the aggregator application using:
+----
+mvn clean package -f aggregator/pom.xml
+----
+
+Setup the necessary environment variables:
+----
+export DEBEZIUM_VERSION=1.4
+export QUARKUS_BUILD=jvm
+----
+
+To launch all the containers and building aggregator container images run the https://github.com/debezium/debezium-examples/blob/master/kstreams-fk-join/docker-compose.yaml[docker-compose.yaml] as:
+----
+docker-compose up --build
+----
+
+Configure the Debezium Connector:
+----
+http PUT http://localhost:8083/connectors/inventory-connector/config < register-postgres.json
+----
+
+Now run an instance of the `debezium/tooling` image, attaching to the same network all the other containers run in:
+----
+docker run --tty --rm \
+    --network kstreams-fk-join-network \
+    debezium/tooling:1.1 \
+----
+This image provides several useful tools such as kafkacat. Within the tooling container, run kafkacat to examine the results of the streaming pipeline:
+
+----
+kafkacat -b kafka:9092 -C -o beginning -q \
+    -t customers-with-addresses | jq .
+----
+
+== Running Natively
+Kafka Streams applications can easily be scaled out i.e. the load will be shared amongst multiple instances of the application, each processing just a subset of the partitions of the input topic(s).
+Running Quarkus applications in native code via GraalVM has a very fast start-up time, the applications use significantly less memory when compiled to native code. This allows you to start as many instances of a Kafka Streams pipeline in parallel in a very memory-efficient way.
+If you want to run this application in `native` mode, set the `QUARKUS_MODE` as `native` and run:
+
+----
+mvn clean package -f aggregator/pom.xml -Pnative
+----
+Here is a https://quarkus.io/guides/kafka-streams#running-natively[detailed guide] for you.
+
+== More Insights on Kafka Streams Extension
+This extension can help you address some of the common requirements when building microservices.
+For running your Kafka Streams application in production, you can also add health checks and metrics for the data pipeline.
+
+https://quarkus.io/guides/microprofile-metrics[Metrics] allow applications to gather various metrics and statistics that provide insights into what is happening inside the application.
+Using the MicroProfile Metrics API, these metrics can be exposed via HTTP using JSON format or the OpenMetrics format.
+From there they can be scraped by tools such as Prometheus and stored for analysis and visualization.
+Apart from application-specific metrics, you can utilize built-in metrics exposed by various Quarkus extensions.
+
+On the other hand, we have https://quarkus.io/guides/microprofile-health[MicroProfile Health], which provides information about the liveness of the application i.e. states whether your application is running or not and whether your application is able to process requests.
```

Comment:
Let's show an example of the health check if one of the input topics is missing.